samcli/commands/remote_invoke/remote_invoke_context.py

```diff
@@ -0,0 +1,122 @@
+    CloudFormationResourceSummary,
+    get_resource_summaries,
+    get_resource_summary_from_physical_id,
+)
+from samcli.lib.utils.resources import AWS_LAMBDA_FUNCTION
+
+LOG = logging.getLogger(__name__)
+
+
+SUPPORTED_SERVICES = {"lambda": AWS_LAMBDA_FUNCTION}
+
+
+class RemoteInvokeContext:
+
+    _boto_client_provider: BotoProviderType
+    _boto_resource_provider: BotoProviderType
+    _stack_name: Optional[str]
+    _resource_id: Optional[str]
+    _resource_summary: Optional[CloudFormationResourceSummary]
+
+    def __init__(
+        self,
+        boto_client_provider: BotoProviderType,
+        boto_resource_provider: BotoProviderType,
+        stack_name: Optional[str],
+        resource_id: Optional[str],
+    ):
+        self._boto_resource_provider = boto_resource_provider
+        self._boto_client_provider = boto_client_provider
+        self._stack_name = stack_name
+        self._resource_id = resource_id
+        self._resource_summary = None
+
+    def __enter__(self) -> "RemoteInvokeContext":
+        self._populate_resource_summary()
+        return self
+
+    def __exit__(self, *args) -> None:
+        pass
+
+    def run(self):
+        pass
+
+    def _populate_resource_summary(self) -> None:
+        if not self._stack_name and not self._resource_id:
+            raise InvalidRemoteInvokeParameters("Either --stack-name or --resource-id parameter should be provided")
+
+        if not self._resource_id:
+            # no resource id provided, list all resources from stack and try to find one
+            self._resource_summary = self._get_single_resource_from_stack()
+            self._resource_id = self._resource_summary.logical_resource_id
+
+        if not self._stack_name:
+            # no stack name provided, resource id should be physical id so that we can use it
+            self._resource_summary = self._get_from_physical_resource_id()
+
+    def _get_single_resource_from_stack(self) -> CloudFormationResourceSummary:
+        """
+        Queries all resources from stack with its type,
+        and returns its information if stack has only one resource from that type (including nested stacks)
+        """
+        resource_summaries = get_resource_summaries(
+            self._boto_resource_provider, self._boto_client_provider, cast(str, self._stack_name), {AWS_LAMBDA_FUNCTION}
+        )
+
+        if len(resource_summaries) == 1:
+            for logical_id, resource_summary in resource_summaries.items():
+                LOG.debug("Using %s resource for remote invocation (%s)", logical_id, resource_summary)
+                return resource_summary
+        elif len(resource_summaries) > 1:
+            raise AmbiguousResourceForRemoteInvoke(
+                f"{self._stack_name} contains more than one resource that could be used with remote invoke, "
+                f"please provide --resource-id to resolve ambiguity."
+            )
+
+        raise NoResourceFoundForRemoteInvoke(
```

Comment:
nit: should we do a `if not resource_summaries` here first.

Reply:
Since we already exhausted all other options above, I was just raising the exception here. If you think that would be useful, I can add a comment on top of it to state this is the case where we don't have any resource summaries at all.